File: src/main/java/com/abrastat/general/Player.java
package com.abrastat.general;

import java.util.HashSet;

import static com.abrastat.general.PlayerBehaviour.BehaviourGroup;

public abstract class Player {

    private String playerName;
    private Pokemon[] pokemonTeam = new Pokemon[6];
    protected boolean justUseFirstAttack = true;
    protected HashSet<PlayerBehaviour> activeBehaviours = new HashSet<>();


    public Player() {

    }

    public void setName(String name)    {
        this.playerName = name;
    }

    public String getName()   {
        return playerName;
    }

    // [0] is always lead
    // [0] will always store the currently battling Pokemon
    public void addPokemon(Pokemon pokemon) {
        this.pokemonTeam[0] = pokemon;
    }

    public Pokemon getPokemon(int partyPosition)    {
        return this.pokemonTeam[partyPosition];
    }

    public abstract Pokemon getCurrentPokemon();

    public HashSet<PlayerBehaviour> getActiveBehaviours()    {
        return this.activeBehaviours;
    }

    public abstract void setBehaviours();


    //TODO eventually...
    public void swapCurrentPokemon()    {

    }

    public Pokemon[] getCurrentTeam() {
        return this.pokemonTeam;
    }

//    public void addPokemon(Pokemon pokemon, int partyPosition)  {
//        this.pokemonTeam[partyPosition] = pokemon;
//    }
//    public void addPokemon(Pokemon p1, Pokemon p2)  {
//        addPokemon(p1, 0);
//        addPokemon(p2, 1);
//    }

    @Override
    public String toString()    {
        return this.playerName;
    }
}

File: src/main/java/com/abrastat/general/Player.kt
package com.abrastat.general

abstract class Player {
    var name: String? = null
    protected val team = arrayOfNulls<Pokemon>(6) // would ArrayList<Pokemon> work better?
    @JvmField
    protected var justUseFirstAttack = true
    @JvmField
    protected var activeBehaviours = HashSet<PlayerBehaviour>()

    abstract val currentPokemon: Pokemon?

    // [0] is always lead
    // [0] will always store the currently battling Pokemon
    fun addPokemon(pokemon: Pokemon) {
        var noRoom = true
        for (i in 0 until 6) {
            if (team[i] == null) {
                team[i] = pokemon
                noRoom = false
                break
            }
        }
        if (noRoom)
            throw IndexOutOfBoundsException(
                    "A team can only contain 6 pokemon"
            )
    }

    fun getPokemon(partyPosition: Int): Pokemon? {
        return team[partyPosition]
    }

    // TODO team having multiple pokemon
    /*
    fun swapCurrentPokemon() {}

    // keep for RBYGame.setWinner
    fun nonFaintedPokemon(): ArrayList<Pokemon> {
        val pokemons = ArrayList<Pokemon>()
        for (i in 0 until 6) {
            if (team[i] != null) {
                if (team[i]!!.nonVolatileStatus != Status.FAINT)
                    pokemons += team[i]!!
            }
        }
        return pokemons
    }
     */
    fun getActiveBehaviours(): HashSet<PlayerBehaviour> {
        if (justUseFirstAttack) {
            activeBehaviours.add(PlayerBehaviour.JUST_ATTACK)
        }
        return activeBehaviours
    }

    abstract fun setBehaviours()

    override fun toString(): String {
        return name!!
    }
}